ns/ref.go

```diff
@@ -0,0 +1,280 @@
+// valid base58 multihashes and makes sure hashes are actually valid base58 multihashes
+// TODO - figure out how IPFS CID's play into this
+func ParseRef(ref string) (Ref, error) {
+	if ref == "" {
+		return Ref{}, ErrEmptyRef
+	}
+
+	var (
+		// nameRefString string
+		dsr = Ref{}
+		err error
+	)
+
+	// if there is an @ symbol, we are dealing with a Ref
+	// with an identifier
+	atIndex := strings.Index(ref, "@")
+
+	if atIndex != -1 {
+
+		dsr.Peername, dsr.Name = parseAlias(ref[:atIndex])
+		dsr.ProfileID, dsr.Path, err = parseIdentifiers(ref[atIndex+1:])
+
+	} else {
+
+		var peername, datasetname, pid bool
+		toks := strings.Split(ref, "/")
+
+		for i, tok := range toks {
+			if isBase58Multihash(tok) {
+				// first hash we encounter is a peerID
+				if !pid {
+					dsr.ProfileID = tok
+					pid = true
+					continue
+				}
+
+				if !isBase58Multihash(toks[i-1]) {
+					dsr.Path = fmt.Sprintf("/%s/%s", toks[i-1], strings.Join(toks[i:], "/"))
+				} else {
+					dsr.Path = fmt.Sprintf("/ipfs/%s", strings.Join(toks[i:], "/"))
+				}
+				break
+			}
+
+			if !peername {
+				dsr.Peername = tok
+				peername = true
+				continue
+			}
+
+			if !datasetname {
+				dsr.Name = tok
+				datasetname = true
+				continue
+			}
+
+			dsr.Path = strings.Join(toks[i:], "/")
+			break
+		}
+	}
+
+	if dsr.ProfileID == "" && dsr.Peername == "" && dsr.Name == "" && dsr.Path == "" {
+		err = fmt.Errorf("malformed Ref string: %s", ref)
+		return dsr, err
+	}
+
+	// if dsr.ProfileID != "" {
+	// 	if !isBase58Multihash(dsr.ProfileID) {
+	// 		err = fmt.Errorf("invalid ProfileID: '%s'", dsr.ProfileID)
+	// 		return dsr, err
+	// 	}
+	// }
+
+	return dsr, err
+}
+
+func parseAlias(alias string) (peer, dataset string) {
+	for i, tok := range strings.Split(alias, "/") {
+		switch i {
+		case 0:
+			peer = tok
+		case 1:
+			dataset = tok
+		}
+	}
+	return
+}
+
+func parseIdentifiers(ids string) (profileID, path string, err error) {
+
+	toks := strings.Split(ids, "/")
+	switch len(toks) {
+	case 0:
+		err = fmt.Errorf("malformed Ref identifier: %s", ids)
+	case 1:
+		if toks[0] != "" {
+			profileID = toks[0]
+
+			return
+		}
+	case 2:
+		profileID = toks[0]
+
+		if isBase58Multihash(toks[0]) && isBase58Multihash(toks[1]) {
+			toks[1] = fmt.Sprintf("/ipfs/%s", toks[1])
+		}
+
+		path = toks[1]
+	default:
+		profileID = toks[0]
+
+		path = fmt.Sprintf("/%s/%s", toks[1], toks[2])
+		return
+	}
+
+	return
+}
+
+// TODO - this could be more robust?
+func stripProtocol(ref string) string {
+	if strings.HasPrefix(ref, "/ipfs/") {
+		return ref[len("/ipfs/"):]
+	}
+	return ref
+}
+
+func isBase58Multihash(hash string) bool {
+	data, err := base58.Decode(hash)
+	if err != nil {
+		return false
+	}
+	if _, err := multihash.Decode(data); err != nil {
+		return false
+	}
+
+	return true
+}
+
+// CompareRef compares two Dataset Refs, returning an error
+// describing any difference between the two references
+func CompareRef(a, b Ref) error {
```

Comment:
A function name starting with "Compare" implies that it returns a boolean, or sometimes a tri-state (like C's strcmp). Instead, can we follow https://github.com/qri-io/qri/blob/2caf757a25a58c8ae1e6e3d7e77e5707fd4fc6bb/rev/rev_test.go#L36 and call this "EnsureRefEqual"? If there are too many external callers, just add a TODO mentioning this.